cpp/tests/experimental/louvain_primitives.cu

```diff
@@ -0,0 +1,197 @@
+#include <experimental/graph.hpp>
+
+#include <patterns/copy_v_transform_reduce_key_aggregated_out_nbr.cuh>
+#include <patterns/transform_reduce_by_adj_matrix_row_col_key_e.cuh>
+#include <utilities/dataframe_buffer.cuh>
+#include <utilities/shuffle_comm.cuh>
+
+#include <rmm/device_uvector.hpp>
+
+template <typename graph_t>
+void update_by_delta_modularity(raft::handle_t const& handle,
+                                graph_t const& graph_view,
+                                std::vector<typename graph_t::vertex_type> cluster_h,
+                                typename graph_t::weight_type total_edge_weight,
+                                typename graph_t::weight_type resolution)
+{
+  using vertex_t = typename graph_t::vertex_type;
+  using edge_t   = typename graph_t::edge_type;
+  using weight_t = typename graph_t::weight_type;
+
+  vertex_t num_verts = graph_view.get_number_of_vertices();
+
+  rmm::device_uvector<vertex_t> cluster_v(num_verts, handle.get_stream());
+
+  raft::update_device(cluster_v.data(), cluster_h.data(), num_verts, handle.get_stream());
+
+  rmm::device_uvector<vertex_t> communities(0, handle.get_stream());
+  rmm::device_uvector<weight_t> community_in_weight_sums(0, handle.get_stream());
+
+  std::tie(communities, community_in_weight_sums) =
+    cugraph::experimental::transform_reduce_by_adj_matrix_col_key_e(
+      handle,
+      graph_view,
+      cluster_v.begin(),
+      cluster_v.begin(),
+      cluster_v.begin(),
+      [] __device__(auto r, auto c, auto w, auto rv, auto cv) {
+        printf("transform reduce (%d,%d,%g,%d,%d)\n", (int)r, (int)c, (float)w, (int)rv, (int)cv);
+        return w;
+      },
+      weight_t{0});
+
+  thrust::for_each(thrust::device,
+                   thrust::make_counting_iterator(0),
+                   thrust::make_counting_iterator(1),
+                   [b_key   = communities.begin(),
+                    e_key   = communities.end(),
+                    b_value = community_in_weight_sums.begin(),
+                    e_value = community_in_weight_sums.end()] __device__(auto) {
+                     char comma[2] = {0, 0};
+                     printf("keys = ");
+                     for (auto p = b_key; p != e_key; ++p) {
+                       printf("%s%d", comma, *p);
+                       comma[0] = ',';
+                     }
+                     printf("\n");
+                     printf("values = ");
+                     comma[0] = (char)0;
+                     for (auto p = b_value; p != e_value; ++p) {
+                       printf("%s%g", comma, *p);
+                       comma[0] = ',';
+                     }
+                     printf("\n");
+                   });
+
+  if (graph_t::is_multi_gpu) {
```

Comment:
This is skipped in my test, as I am single GPU.  However, isn't this going to have the same problem as my above example?  That is, if vertex u is distributed over multiple GPUs, and generates key/value pairs with the same key on different GPUs, this step will only migrate the key/value pairs to the correct GPU.  No reduction will take place here.

Reply:
The behavior above is clearly a bug. Values for the same keys are globally reduced. And a part of the final (key, value) pairs is returned in every GPU.

Reply:
I haven't tested this.  I was speculating here.